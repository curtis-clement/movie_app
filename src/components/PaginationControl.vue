<script setup lang="ts">
import { PaginationDirection } from '@/models/helpers.model';

const props = defineProps<{
  currentPage: number;
  hasNextPage: boolean;
}>();

const emit = defineEmits(['page-change']);

function changePage(direction: PaginationDirection) {
  emit('page-change', direction === PaginationDirection.NEXT ? props.currentPage + 1 : props.currentPage - 1);
}
</script>

<template>
  <div class="pagination-control">
    <button 
      class="pagination-button"
      :disabled="currentPage === 1"
      @click="changePage(PaginationDirection.PREVIOUS)"
    >
      &lt;
    </button>
    <span class="page-indicator">Page {{ currentPage }}</span>
    <button 
      class="pagination-button"
      :disabled="!hasNextPage"
      @click="changePage(PaginationDirection.NEXT)"
    >
      &gt;
    </button>
  </div>
</template>

<style scoped>
.pagination-control {
  align-items: center;
  display: flex;
  gap: 1rem;
  justify-content: center;
  margin: 2rem 0;
}

.pagination-button {
  background-color: white;
  border: 1px solid #3498db;
  border-radius: 4px;
  color: #3498db;
  cursor: pointer;
  padding: 0.5rem 1rem;
  transition: background-color 0.2s;
}

.pagination-button:hover:not(:disabled) {
  background-color: #3498db;
  color: white;
}

.pagination-button:disabled {
  cursor: not-allowed;
  opacity: 0.5;
}

.page-indicator {
  font-weight: 600;
}
</style>
