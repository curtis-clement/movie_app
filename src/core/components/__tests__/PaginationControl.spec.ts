import { describe, it, expect, beforeEach } from 'vitest';
import { mount, VueWrapper } from '@vue/test-utils';
import PaginationControl from '@/core/components/PaginationControl.vue';

const mountComponent = (props: {
  currentPage: number;
  hasNextPage: boolean;
}) => {
  return mount(PaginationControl, {
    props,
  });
};

describe('PaginationControl', () => {
  let wrapper: VueWrapper<InstanceType<typeof PaginationControl>>;

  beforeEach(() => {
    wrapper = mountComponent({
      currentPage: 1,
      hasNextPage: true,
    });
  }); 

  it('renders correctly with initial props', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('passes props correctly', () => {
    expect(wrapper.props()).toEqual({
      currentPage: 1,
      hasNextPage: true,
    });
  });

  it('disables previous button when on first page', () => {
    wrapper.setProps({ currentPage: 1 });
    const prevButton = wrapper.findAll('.pagination-button')[0];
    expect(prevButton.attributes('disabled')).toBeDefined();
  });

  it('enables previous button when not on first page', async () => {
    wrapper.setProps({ currentPage: 2 });
    await wrapper.vm.$nextTick();
    const prevButton = wrapper.findAll('.pagination-button')[0];
    expect(prevButton.attributes('disabled')).toBeUndefined();
  });

  it('disables next button when hasNextPage is false', async () => {
    wrapper.setProps({ hasNextPage: false });
    await wrapper.vm.$nextTick();
    const nextButton = wrapper.findAll('.pagination-button')[1];
    expect(nextButton.attributes('disabled')).toBeDefined();
  });

  it('enables next button when hasNextPage is true', async () => {
    const nextButton = wrapper.findAll('.pagination-button')[1];
    expect(nextButton.attributes('disabled')).toBeUndefined();
  });

  it('emits page-change event with correct value when clicking next', async () => {    
    const nextButton = wrapper.findAll('.pagination-button')[1];
    await nextButton.trigger('click');

    expect(wrapper.emitted('page-change')).toBeTruthy();
    expect(wrapper.emitted('page-change')?.[0]).toEqual([2]);
  });

  it('emits page-change event with correct value when clicking previous', async () => {
    wrapper.setProps({ currentPage: 2 });
    await wrapper.vm.$nextTick();
    const prevButton = wrapper.findAll('.pagination-button')[0];
    await prevButton.trigger('click');

    expect(wrapper.emitted('page-change')).toBeTruthy();
    expect(wrapper.emitted('page-change')?.[0]).toEqual([1]);
  });

  it('displays correct page number', async () => {
    wrapper.setProps({ currentPage: 5 });
    await wrapper.vm.$nextTick();
    expect(wrapper.find('.page-indicator').text()).toBe('Page 5');
  });
});