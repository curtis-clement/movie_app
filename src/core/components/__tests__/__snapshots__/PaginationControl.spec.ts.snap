// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PaginationControl > renders correctly with initial props 1`] = `"<div data-v-b94b60b0="" class="pagination-control"><button data-v-b94b60b0="" class="pagination-button" disabled=""> &lt; </button><span data-v-b94b60b0="" class="page-indicator">Page 1</span><button data-v-b94b60b0="" class="pagination-button"> &gt; </button></div>"`;
